### DISCUSSION TITLE
how can I get access to the URL on server side when using app dir?

### DISCUSSION THREAD

**Discussion by JacobWeisenburger** [vercel/next.js#47280] [Help] [answered]
### Summary

how can I get access to the URL on server side when using app dir?

**Comment by icyJoseph** [comment]
Half of this answer is that during pre-rendering there's no browsing going on, so part of the rendering process has to account for lack of a "URL". At best you know the path your are being rendered at.

Then we have the on-demand rendering scenarios, like for example: GSSP... even there I think you can only get the resolved path, but the full URL is not known... IIRC this is because there might be another layer handling traffic (reverse proxy). Vercel exposes the VERCEL_URL env var for cases where you really need the URL.

**Reply by JacobWeisenburger** [reply]
How am I supposed to access `VERCEL_URL`?

```tsx
export default async function Page () {
    console.log( process.env.VERCEL_URL ) // undefined
    return <div></div>
}
```

**Reply by icyJoseph** [reply]
You need to activate it on the dashboard, https://vercel.com/docs/concepts/projects/environment-variables/system-environment-variables

**Reply by JacobWeisenburger** [reply]
I'm not using Vercel to host this project. So.....

**Reply by icyJoseph** [reply]
Check for the host header then? There's a headers function in server components https://beta.nextjs.org/docs/api-reference/headers

**Reply by JacobWeisenburger** [reply]
that only gives me the host. I am trying to get the full url.

**Reply by JacobWeisenburger** [reply]
> At best you know the path your are being rendered at.

@icyJoseph, how do you get access to the path in a server side component?

**Comment by JacobWeisenburger** [comment]
It seems weird to me that there isn't an easy way to do something as basic as get the url.

---

In deno fresh, the url is provided as a prop to every page. https://fresh.deno.dev/docs/concepts/routes#component-route.

```tsx
import { PageProps } from "$fresh/server.ts";

export default function Page(props: PageProps) {
  return <div>You are on the page '{props.url.href}'.</div>;
}
```

---
Astro also makes this really easy to access the url.
https://docs.astro.build/en/reference/api-reference/#astrourl

```tsx
<h1>The current URL is: {Astro.url}</h1>
<h1>The current URL pathname is: {Astro.url.pathname}</h1>
<h1>The current URL origin is: {Astro.url.origin}</h1>
```
---


Am I just not understanding something?

**Comment by apecollector** [comment]
I don't think so, the best you can do is use a client hook

**Reply by JacobWeisenburger** [reply]
Thank you for your answer.

This seems really weird to me to not give easy access to the URL or even just the pathname on the server side, considering that it's so easy in other frameworks.

**Comment by radosek** [comment]
```js
import { NextRequest } from "next/server"
import { headers } from "next/headers"

const headersList = headers()
const referer = headersList.get("referer")

if (referer) {
  const request = new NextRequest(referer)
  console.log(request.nextUrl.pathname)
}
```

**Reply by radosek** [reply]
> ![image](https://user-images.githubusercontent.com/31667350/230213713-a1f397df-8ebf-410e-ab1c-bd2166458db4.png)
> 
> 
> 
> 

You need to use it inside the function that renders/return children component (or any other component that wraps children.

![image](https://user-images.githubusercontent.com/47051193/230214835-c5de2e54-3570-484a-9723-1673be76f7a1.jpeg)

**Reply by JacobWeisenburger** [reply]
Thanks. It's working now, except that because it's using `referer` it's giving me the correct url less than 25% of the time.

![image](https://user-images.githubusercontent.com/31667350/230215477-b65a313d-98f7-4b1d-a251-14e3491d77e0.png)

![image](https://user-images.githubusercontent.com/31667350/230215513-07e61fe2-bb86-461c-8d31-91848248f002.png)

**Reply by radosek** [reply]
https://beta.nextjs.org/docs/api-reference/headers

Check other headers that comes with a request too, but you might need to parse the URL manually.

**Reply by JacobWeisenburger** [reply]
```ts
{
    'accept': '*/*',
    'accept-encoding': 'gzip, deflate, br',
    'accept-language': 'en-US,en;q=0.9,zh-CN;q=0.8,zh;q=0.7,es;q=0.6',
    'connection': 'keep-alive',
    'dnt': '1',
    'host': 'localhost:3000',
    'referer': 'http://localhost:3000/',
    'sec-ch-ua': '"Chromium";v="112", "Google Chrome";v="112", "Not:A-Brand";v="99"',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '"Windows"',
    'sec-fetch-dest': 'empty',
    'sec-fetch-mode': 'cors',
    'sec-fetch-site': 'same-origin',
    'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36'
}
```

the only one that is close to a url here is `referer`, except that it's wrong most of the time.

**Reply by radosek** [reply]
https://github.com/vercel/next.js/issues/43704

**Reply by JacobWeisenburger** [reply]
should I be looking at a specific part of that thread? or are you just telling me to read the whole thing?

**Reply by radosek** [reply]
Did not tried it out but there are two solutions:

1) setting own header inside middleware (from `request.nextUrl.pathname`) and then getting it in `layout.tsx` with headers()

2) getting it from `children.props.childProp.segment`

**Reply by JacobWeisenburger** [reply]
https://github.com/vercel/next.js/issues/46618#issuecomment-1450416633

This ended up working for me. Thanks for your patience and tenacity.

**Comment by JacobWeisenburger** [comment, marked as answer]
https://github.com/vercel/next.js/issues/46618#issuecomment-1450416633

**Comment by MESSELMIyahya** [comment]
If you're using the i18n in the middleware do this :
```ts
import { NextRequest } from "next/server";
import createMiddleware from "next-intl/middleware";
import { routing } from "./i18n/routing";

const intlMiddleware = createMiddleware(routing);

export function middleware(req: NextRequest) {
  const url = req.nextUrl.clone();
  const originalUrl = url.pathname + url.search;

  // Set the x-original-url header
  req.headers.set("x-original-url", originalUrl);

  // Call the existing intl middleware
  return intlMiddleware(req);
}

export const config = {
  // Match only internationalized pathnames
  matcher: ["/", "/(ar|fr|en)/:path*"],
};
```

and then you can access it in your server side component like this 
```ts
 headers().get('x-original-url') 
```
